### PR TITLE
WASI prototype design, implementation, and documentation.

### PR DESCRIPTION
**Pull request by sunfishcode**
This adds documents describing the WASI Core API, and an implementation in
Wasmtime.

For more information about WASI, see
  https://hacks.mozilla.org/2019/03/standardizing-wasi-a-webassembly-system-interface/

For documentation, see the docs directory in this PR, which is also linked to from
  https://wasi.dev.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,26 +14,27 @@ utility or as a library embedded in a larger application.
...
-*Wasmtime is complete enough to pass the WebAssembly spec testsuite.* Support for
-system APIs is coming soon!
+Wasmtime passes the WebAssembly spec testsuite, and supports a new system
+API proposal called [WebAssembly System Interface], or WASI.
 
-One goal for this project is to implement [CloudABI](https://cloudabi.org/) using
-WebAssembly as the code format, provide [CloudABI system calls] as WebAssembly
-host imports, and then port the [Rust CloudABI package] and [CloudABI libc] to it
-to support Rust, C, C++, and other toolchains.
+There are Rust C, and C++ toolchains that can compile programs with WASI. See
```
Comment (MSleepyPanda) [inline README.md]:
```suggestion
There are Rust, C, and C++ toolchains that can compile programs with WASI. See
```

Comment (sunfishcode) [reply README.md]: Indeed, thanks!

**Comment by maxmcd** [comment]
Whoaaa nice!

File: src/wasmtime.rs [hunk 2]
```diff
@@ -59,26 +63,36 @@ including calling the start function if one is present. Additional functions
...
+    wasmtime [-odg] [--preload=<wasm>...] [--env=<env>...] [--dir=<dir>...] [--mapdir=<mapping>...] <file> [<arg>...]
+    wasmtime [-odg] [--preload=<wasm>...] [--env=<env>...] [--dir=<dir>...] [--mapdir=<mapping>...] --invoke=<fn> <file> [<arg>...]
     wasmtime --help | --version
 
 Options:
     --invoke=<fn>       name of function to run
     -o, --optimize      runs optimization passes on the translated functions
     -g                  generate debug information
     -d, --debug         enable debug output on stderr/stdout
+    --preload=<wasm>    load an additional wasm module before loading the main module
```
Comment (bjorn3) [inline src/wasmtime.rs]: Maybe allow multiple modules to be preloaded?

File: src/wasmtime.rs [hunk 3]
```diff
@@ -100,6 +114,91 @@ fn read_wasm(path: PathBuf) -> Result<Vec<u8>, String> {
...
+fn compute_argv(argv0: &str, arg_arg: &[String]) -> Vec<String> {
+    let mut result = Vec::new();
+
+    // Add argv[0], which is the program name. Only include the base name of the
+    // main wasm module, to avoid leaking path information.
+    result.push(
+        Path::new(argv0)
+            .components()
+            .next_back()
+            .map(Component::as_os_str)
```
Comment (bjorn3) [inline src/wasmtime.rs]: You can use `path.file_name()`: https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.file_name

File: src/wasmtime.rs [hunk 4]
```diff
@@ -139,20 +238,52 @@ fn main() {
...
+    let argv = compute_argv(&args.arg_file, &args.arg_arg);
+    let environ = compute_environ(&args.flag_env);
+    context.name_instance(
+        "wasi_unstable".to_owned(),
+        instantiate_wasi("", global_exports, &preopen_dirs, &argv, &environ)
+            .expect("instantiating wasi"),
+    );
+
+    // FIXME: Also recognize "env", for compatibility with clang/llvm 8.0. And use
+    // "__wasi_" prefixes for compaitility with prototype reference-sysroot.
```
Comment (bjorn3) [inline src/wasmtime.rs]:
```suggestion
    // "__wasi_" prefixes for compatibility with prototype reference-sysroot.
```

File: src/wasmtime.rs (hunk 2 above)
Comment (sunfishcode) [reply src/wasmtime.rs]: You can use --preload= multiple times.

File: src/wasmtime.rs (hunk 3 above)
Comment (sunfishcode) [reply src/wasmtime.rs]: Ah, I was looking for a function that did that, but somehow missed it. Thanks!

**Comment by sunfishcode** [comment]
Thanks for the reviews all! I'm going to merge this now to make it easier for people to experiment with WASI, but this is just the beginning :smile: .

File: src/wasmtime.rs (hunk 2 above)
Comment (bjorn3) [reply src/wasmtime.rs]: Didn't know that.

**Pull request merged**